File: src/gestorAplicacion/administracion/Grupo.java
package gestorAplicacion.administracion;
import java.io.Serializable;
import java.util.*;
import gestorAplicacion.usuario.*;

public class Grupo implements Serializable{
	private static final long serialVersionUID = 1L;
	private Materia materia;
    private int numero;
    private Profesor profesor;
    private ArrayList<String> horario; 
    private int cupos;
    private Salon salon;
    private ArrayList<Estudiante> estudiantes;
	private static ArrayList<Grupo> gruposTotales = new ArrayList<Grupo>();

    public Grupo(Materia materia, int numero, Profesor profesor, ArrayList<String> horario, int cupos, Salon salon){
      this.materia = materia;
      this.numero=numero;
      this.profesor=profesor;
      this.horario=horario;
      this.cupos=cupos;
      this.salon=salon;
      this.estudiantes=new ArrayList<Estudiante>();
	  Grupo.gruposTotales.add(this);
    } 
    
	public Grupo(Materia materia, int numero, Profesor profesor){
      this.materia = materia;
      this.numero=numero;
      this.profesor=profesor;
	  Grupo.gruposTotales.add(this);
    } 

    public String mostrarInformacionGrupo(){
			String retorno= "Número del grupo: "+this.numero+", Profesor: "+this.profesor+", Horario: "+this.horario+", Cupos: "+this.cupos+", Salón: "+this.salon;
			return retorno;
		}


    public boolean existenciaEstudiante(Estudiante estudiante){
      for (int j=0;j<this.estudiantes.size();j++){
        if (this.estudiantes.get(j).getId() == estudiante.getId()){
          return true;
        }
      }
      return false;
    }

	public void eliminarEstudiante(Estudiante estudiante){
		int indice = -1;
		for (int i=0;i<this.estudiantes.size();i++){
			if (this.estudiantes.get(i).getNombre().equals(estudiante.getNombre())){
					indice = i;
					this.cupos++;
					estudiante.eliminarGrupo(this);
					break;
			}
		}
		if (indice != -1){
			this.estudiantes.remove(indice);
		}
    }

	public static Grupo buscarGrupo(Materia materiaE, Grupo grupoE){
		int indicei = -1;
		int indicej = -1;
		for (int i = 0; i<Materia.getMateriasTotales().size(); i++){
			Materia materia = Materia.getMateriasTotales().get(i);
			if (materia.getNombre().equals(materiaE.getNombre())){
				indicei = i;
				for (int j = 0; j<materia.getGrupos().size(); j++){
					Grupo grupo = materia.getGrupos().get(j);
					if (grupo.getNumero() == grupoE.getNumero()){
						indicej = j;
						break;
					}
				}
			}
		}
		return Materia.getMateriasTotales().get(indicei).getGrupos().get(indicej);
	}

    public void agregarEstudiante(Estudiante estudiante){
      this.estudiantes.add(estudiante);
	  	this.cupos--;
    }

	public int getNumero() {
		return this.numero;
	}

	public void setNumero(int numero) {
		this.numero = numero;
	}

	public Profesor getProfesor() {
		return this.profesor;
	}

	public void setProfesor(Profesor profesor) {
		this.profesor = profesor;
	}

	public ArrayList<String> getHorario() {
		return this.horario;
	}

	public void setHorario(ArrayList<String> horario) {
		this.horario = horario;
	}

	public int getCupos() {
		return this.cupos;
	}

	public void setCupos(int cupos) {
		this.cupos = cupos;
	}

	public Salon getSalon() {
		return this.salon;
	}

	public void setSalon(Salon salon) {
		this.salon = salon;
	}

	public ArrayList<Estudiante> getEstudiantes() {
		return this.estudiantes;
	}

	public void setEstudiantes(ArrayList<Estudiante> estudiantes) {
		this.estudiantes = estudiantes;
	}
	
	public Materia getMateria() {
		return this.materia;
	}
	
	public void setMateria(Materia materia) {
		this.materia = materia;
	}

	public static ArrayList<Grupo> getGruposTotales() {
		return gruposTotales;
	}
	
}

File: Python/src/gestorAplicacion/administracion/Grupo.py
# Autores:
#  MATEO ÁLVAREZ MURILLO
#  EFRAÍN GÓMEZ RAMÍREZ
#  LIBARDO JOSÉ NAVARRO PEDROZO
#  SEBASTIÁN OCAMPO GALVIS
# from gestorAplicacion.usuario.Estudiante import Estudiante
# from Salon import Salon


class Grupo:
    _gruposTotales = []

    def __init__(self, materia, numero, profesor, horario=[], cupos=None, salon=None):
        self._materia = materia
        self._numero = numero
        self._profesor = profesor
        self._horario = horario if horario is not None else []
        self._cupos = cupos
        self._salon = salon
        self._estudiantes = []
        Grupo._gruposTotales.append(self)

    def mostrarInformacionGrupo(self):
        return "Número del grupo: {}, Profesor: {}, Horario: {}, Cupos: {}, Salón: {}".format(
            self.numero, self.profesor, self.horario, self.cupos, self.salon
        )

    def existenciaEstudiante(self, estudiante):
        for e in self._estudiantes:
            if e.getId() == estudiante.getId():
                return True
        return False

    def eliminarEstudiante(self, estudiante):
        indice = -1
        for i, e in enumerate(self._estudiantes):
            if e.getNombre() == estudiante.getNombre():
                indice = i
                self._cupos += 1
                estudiante.eliminarGrupo(self)
                break
        if indice != -1:
            self._estudiantes.pop(indice)

    @staticmethod
    def buscarGrupo(materiaE, grupoE):
        from gestorAplicacion.administracion.Materia import Materia
        
        indicei = -1
        indicej = -1
        for i in range(len(Materia.getMateriasTotales())):
            materia = Materia.getMateriasTotales()[i]
            if materia.getNombre() == materiaE.getNombre():
                indicei = i
                for j in range(len(materia.getGrupos())):
                    grupo = materia.getGrupos()[j]
                    if grupo.getNumero() == grupoE.getNumero():
                        indicej = j
                        break

        return Materia.getMateriasTotales()[indicei].getGrupos()[indicej]

    def agregarEstudiante(self, estudiante):
        self._estudiantes.append(estudiante)
        self._cupos -= 1

    def getNumero(self):
        return self._numero

    def setNumero(self, numero):
        self._numero = numero

    def getProfesor(self):
        return self._profesor

    def setProfesor(self, profesor):
        self._profesor = profesor

    def getHorario(self):
        return self._horario

    def setHorario(self, horario):
        self._horario = horario

    def getCupos(self):
        return self._cupos

    def setCupos(self, cupos):
        self._cupos = cupos

    def getSalon(self):
        return self._salon

    def setSalon(self, salon):
        self._salon = salon

    def getEstudiantes(self):
        return self._estudiantes

    def setEstudiantes(self, estudiantes):
        self._estudiantes = estudiantes

    def getMateria(self):
        return self._materia

    def setMateria(self, materia):
        self._materia = materia

    @classmethod
    def getGruposTotales(cls):
        return cls._gruposTotales

    @classmethod
    def setGruposTotales(cls, grupos):
        cls._gruposTotales = grupos
